### PR TITLE
[WIP] Support Async lookup for Paimon-flink

### PR DESCRIPTION
**Pull request by Myasuka**
### Purpose

Fix https://github.com/apache/incubator-paimon/issues/848

### Tests

Added `LookupJoinITCase` for async lookup case.

### API and Format 

No new API or format.

### Documentation

Will add documentation later.

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/FileStoreLookupFunction.java [hunk 1]
```diff
@@ -120,10 +139,20 @@ public FileStoreLookupFunction(
...
     public void open(FunctionContext context) throws Exception {
-        String tmpDirectory = getTmpDirectory(context);
+        Options options = Options.fromMap(table.options());
+
+        this.lookupAsync = options.get(FlinkConnectorOptions.LOOKUP_ASYNC);
+        if (lookupAsync) {
+            this.lock = new ReentrantLock();
+            this.asyncThreadPool =
+                    Executors.newFixedThreadPool(
+                            ASYNC_THREAD_POOL_SIZE,
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/FileStoreLookupFunction.java:149]: Actually, `FileStoreLookupFunction` is not a thread safe class.
You can take a look to `RocksDBState.serializeKey`.

So the thread number should alway be 1 only.

**Review by JingsongLi — COMMENTED** [review]
Thanks @Myasuka for the contribution!

Can we just have a wrapper for `FileStoreLookupFunction`?
Maybe we can introduce an interface `LookupFunction`.
```
void open(FunctionContext context) throws Exception;
Collection<RowData> lookup(RowData keyRow);
void close() throws IOException;
```

And we can provide an `AsyncLookupFunction` to wrap `FileStoreLookupFunction`.

In this way, we can keep `FileStoreLookupFunction` clean.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/FileStoreLookupFunction.java (hunk 1 above)
Comment (Myasuka) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/FileStoreLookupFunction.java:149]: Got it, thanks for the reminder.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkConnectorOptions.java [hunk 2]
```diff
@@ -181,6 +181,12 @@ public class FlinkConnectorOptions {
                             "How many splits should assign to subtask per batch in StaticFileStoreSplitEnumerator "
                                     + "to avoid exceed `akka.framesize` limit.");
 
+    public static final ConfigOption<Boolean> LOOKUP_ASYNC =
```
[the remark this answers is not in the document]
Comment (Myasuka) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkConnectorOptions.java:184]: I feel a bit confused about this feature. From the FLIP and docs, we can see the join-hint will prefer async lookup join when we add hint `LOOKUP('table'='dim1', 'async'='true')`. 
However, if we take a look at the implementation:
[LookupJoinUtil#isAsyncLookup](https://github.com/apache/flink/blob/678370b18e1b6c4a23e5ce08f8efd05675a0cc17/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/LookupJoinUtil.java#L412-L413), it will return async enabled via judging whether the created `lookupRuntimeProvider` is an instance of `AsyncLookupFunctionProvider`.
However we would only [return the sync lookup](https://github.com/apache/flink/blob/678370b18e1b6c4a23e5ce08f8efd05675a0cc17/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/LookupJoinUtil.java#L614-L626) runtime provider without async option in connector DDL. 
In other words, if we did not give async related options in the connector DDL, the `LOOKUP('async'='true')` hint would not take effect.

[the remark this answers is not in the document]
Comment (JingsongLi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkConnectorOptions.java:184]: CC @lincoln-lil

[the remark this answers is not in the document]
Comment (lincoln-lil) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkConnectorOptions.java:184]: @Myasuka @JingsongLi  sorry for the late response. For flink planner, it does prefer the async lookup and will check what kind of lookup function does the backend source can provide 
```
// prefer (not require) by default
boolean preferAsync = preferAsync(lookupHint);
// check lookup function provider ...
return preferAsync ? asyncFound : !syncFound;
```
so, if the connector can have both sync and async lookup capabilities, it‘s better to implement both providers

[the remark this answers is not in the document]
Comment (Myasuka) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkConnectorOptions.java:184]: @lincoln-lil The `preferAsync` still need the `asyncFound` to return `true` for async-lookup. That is to say, we still need the instantiated provider is an `AsyncLookupFunctionProvider`. If so, we have to introduce anther connector-scope parameter of `async-lookup` to instantiate the provider. On the other hand, @JingsongLi thought we could leverage the lookup hint instead of introducing another parameter. That's the point to confuse me.

[the remark this answers is not in the document]
Comment (lincoln-lil) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkConnectorOptions.java:184]: @Myasuka I see your confusion. With the current table interfaces, the connector itself needs to provide a `LookupRuntimeProvider` that is both an `AsyncLookupRuntimeProvider` and a `LookupFunctionProvider` to achieve this goal, like this sample code:
```java
/** A full capability lookup table source. */
public static class MixedLookupTableSource implements LookupTableSource {

        @Override
        public LookupRuntimeProvider getLookupRuntimeProvider(LookupContext context) {
            AsyncLookupFunction asyncLookupFunction = new MyAsyncLookupFunction...
            LookupFunction lookupFunction = new MyLookupFunction...
            return MixedLookupFunctionProvider.of(asyncLookupFunction, lookupFunction);
        }

        interface MixedLookupFunctionProvider
                extends AsyncLookupFunctionProvider, LookupFunctionProvider {
            static MixedLookupFunctionProvider of(AsyncLookupFunction asyncLookupFunction, LookupFunction lookupFunction) {
                return new MixedLookupFunctionProvider() {
                    @Override
                    public AsyncLookupFunction createAsyncLookupFunction() {
                        return asyncLookupFunction;
                    }

                    @Override
                    public LookupFunction createLookupFunction() {
                        return lookupFunction;
                    }
                };
            }
        }
```
Yes, for now it looks like this part of the code might be easier for the connector developer if the planner can figure out how to handle it, such as adding a boolean preferAsync() to the `LookupContext`, which would also save the overhead of creating a new function, I'll track this in flink.

**Comment by JingsongLi** [comment]
I have created a new implementation in https://github.com/apache/incubator-paimon/pull/1770

**Comment by Myasuka** [comment]
Close this PR as #1770 merged.
